lib/rouge/lexers/ghc_cmm.rb

```diff
@@ -0,0 +1,337 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+# C minus minus (Cmm) is a pun on the name C++. It's an intermediate language
+# of the Glasgow Haskell Compiler (GHC) that is very similar to C, but with
+# many features missing and some special constructs.
+#
+# Cmm is a dialect of C--. The goal of this lexer is to use what GHC produces
+# and parses (Cmm); C-- itself is not supported.
+#
+# https://gitlab.haskell.org/ghc/ghc/wikis/commentary/compiler/cmm-syntax
+#
+module Rouge
+  module Lexers
+    class GHCCmm < RegexLexer
+      title "GHC Cmm (C--)"
+      desc "GHC Cmm is the intermediate representation of the GHC Haskell compiler"
+      tag 'ghc-cmm'
+      aliases 'cmm'
+      filenames '*.cmm', '*.dump-cmm', '*.dump-cmm-*'
+
+      macros = %r(#(?:define|include|endif|else|if))
+
+      ws = %r(\s|//.*?\n|/[*](?:[^*]|(?:[*][^/]))*[*]+/)mx
+      id = %r((?!#{macros})[\w#\$%_']+)
+
+      complex_id = %r(
+        (?:[\w#$%_']|\(\)|\(,\)|\[\]|[0-9])*
+        (?:[\w#$%_']+)
+      )mx
+
+      state :root do
+        rule %r/\s+/m, Text
+
+        # sections markers
+        rule %r/^=====.*=====$/, Generic::Heading
+
+        # timestamps
+        rule %r/^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d+ UTC$/, Comment::Single
+
+        mixin :detect_section
+        mixin :preprocessor_macros
+
+        mixin :info_tbls
+        mixin :comments
+        mixin :literals
+        mixin :keywords
+        mixin :types
+        mixin :infos
+        mixin :names
+        mixin :operators
+
+        # escaped newline
+        rule %r/\\\n/, Comment::Preproc
+
+        # rest is Text
+        rule %r/./, Text
+      end
+
+      state :detect_section do
+        rule %r/(section)(\s+)/ do |m|
+          token Keyword, m[1]
+          token Text, m[2]
+          push :section
+        end
+      end
+
+      state :section do
+        rule %r/"(data|cstring|text|rodata|relrodata|bss)"/, Name::Builtin
+
+        rule %r/{/, Punctuation, :pop!
+
+        mixin :names
+        mixin :operators
+        mixin :keywords
+
+        rule %r/\s/, Text
+      end
+
+      state :preprocessor_macros do
+        rule %r/#(include|endif|else|if)/, Comment::Preproc
+
+        rule %r{
+            (\#define)
+            (#{ws}*)
+            (#{id})
+          }mx do |m|
+          token Comment::Preproc, m[1]
+          recurse m[2]
```

Comment:
Sorry, this might be a silly question but why do you need to recurse on the whitespace?

Reply:
I think because you could write:

```c
#define /* comment_1 */ /* comment_2 */ foo bar
```

That comments can arise almost everywhere is one aspect that made this lexer complicated.